acceptance: introduce partition nemesis in acceptance tests

tbg opened the pull request:
```
This nemesis allows node partitions to be set up during tests. Partitions can
be bidirectional, but also one-way (i.e. one node can initiate connections to
another, but not the other way around). The latter likely isn't too
interesting with the change to streaming RPCs, but it should be possible to
introduce interesting phenomena nevertheless.

Wrote a bank test with the nemesis. I found the existing bank test too
complicated for its own sake, so I re-wrote it. The plan is to make it
the more advanced bank example with constraints which would fail under
SNAPSHOT isolation (i.e. transferring money from A and B to C and D under
the constraint that A+B >= 0).
```

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4712)

<!-- Reviewable:end -->

mrtracy commented:
Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[acceptance/partition.go, line 59 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBiEYUxMrjdiO4ve1z_:-KBiEYUxMrjdiO4ve1za:-407379893) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/partition.go#L59)):*
It appears that this isn't just a one-time random sever; it appears to randomly sever the network into two groups *and then restore it*, several times as long as the nemesis is running.  If that's the case, the comment should be extended.

---

*[acceptance/partition.go, line 99 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBiFHyXd6_isbLVzrzD:-KBiFHyXd6_isbLVzrzE:-514470471) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/partition.go#L99)):*
Does this need to be in a goroutine?  The only caller immediately blocks on the returned channel.

---

*[acceptance/iptables/iptables.go, line 18 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBiDOZKwJ8jvP-JO5Oa:-KBiDOZKwJ8jvP-JO5Ob:344914360) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/iptables/iptables.go#L18)):*
Should `blacklist` be a type? Also, maybe this method should be "BidirectionalBlacklist", because it seems that iptables command can do more than blacklist IPs.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 11 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[acceptance/partition.go, line 59 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBiEYUxMrjdiO4ve1z_:-KByeFiwLNdNZiCmLIeC:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/partition.go#L59)):*
Done

---

*[acceptance/partition.go, line 99 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBiFHyXd6_isbLVzrzD:-KByg1Q2OUOa81_8oukZ:2074308490) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/partition.go#L99)):*
No, doesn't seem necessary, at least not currently. Removed.

---

*[acceptance/iptables/iptables.go, line 18 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBiDOZKwJ8jvP-JO5Oa:-KBygD5N1hagSuULR8Qp:-921688077) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/iptables/iptables.go#L18)):*
Yeah, it could be a type. I'd like to leave as is for now since the blacklist is mostly relevant for the lowest level API and likely won't be called directly from tests. Same for renaming, we can do that when we actually use iptables for other stuff, which may never happen.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Looks like this just broke with Docker 1.10 since privileged mode is allowed no more. I opened a support request with CircleCI to see whether anything can be done about that.

tschottdorf commented:
Resuscitated! PTAL. Doesn't run on CircleCI but fit for nightlies and local testing. I'm planning to make this work against GCE/AWS (or at least try to), which shouldn't be so bad.

tamird commented:
Reviewed 4 of 11 files at r1, 1 of 5 files at r2, 6 of 10 files at r5, 2 of 2 files at r7.
Review status: 8 of 10 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---

*[acceptance/chaos_test.go, line 307 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPo9FKBBp92s7NqPTo:-KFPo9FKBBp92s7NqPTp:166874) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/chaos_test.go#L307)):*
?

---

*[acceptance/cluster/cluster.go, line 39 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqBoj8QrcMQlpDY6q:-KFPqBoj8QrcMQlpDY6r:619753737) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/cluster.go#L39)):*
shouldn't this return a `net.Addr`? also looks like it's only used to get the hostname, so maybe just return that.

---

*[acceptance/cluster/localcluster.go, line 36 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPpYichEFmZFc7HrRc:-KFPpYichEFmZFc7HrRd:-1001917522) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L36)):*
could we avoid depending on docker/docker?

---

*[acceptance/cluster/localcluster.go, line 384 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPpq-SbEXIRs46Nxmx:-KFPpq-SbEXIRs46Nxmy:1762091086) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L384)):*
should `l.networkID` be of type `container.NetworkMode`?

---

*[acceptance/cluster/localcluster.go, line 737 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqR6ITDtz0izh-4zr:-KFPqR6ITDtz0izh-4zs:1072445900) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L737)):*
I think you should use a `util.UnresolvedAddr` here to avoid needing `defaultPort`.

---

*[acceptance/cluster/localcluster.go, line 811 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqkSWBitVRwCbDEi0:-KFPqkSWBitVRwCbDEi1:743849634) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L811)):*
nit: do this right after the first error check

---

*[acceptance/cluster/localcluster.go, line 826 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqwkpuakNZq7UwZni:-KFPqwkpuakNZq7UwZnj:1738779817) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L826)):*
nit: dubious use of strings.Join - default stringer is probably ok (or use %v which will just wrap with `[]`)

---

*[acceptance/iptables/iptables.go, line 15 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPrDRCAv1Wck5nf-O2:-KFPrDRCAv1Wck5nf-O3:-1281148122) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/iptables/iptables.go#L15)):*
reflow

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 8 of 10 files at r5, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[acceptance/partition_test.go, line 1 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFQWeN4Re22KBzr56ay:-KFQWeN4Re22KBzr56az:1225025578) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/partition_test.go#L1)):*
Some of the new files in this PR are missing license headers.

---

*[acceptance/iptables/iptables.go, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBip_uwPItZGyzD_J7E:-KBip_uwPItZGyzD_J7F:329819358) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/iptables/iptables.go#L15)):*
Wrap this comment at 80 chars.

---

*[acceptance/iptables/iptables.go, line 34 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFQXrd2TeTVrvJ_PKUO:-KFQXrd2TeTVrvJ_PKUP:494370170) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/iptables/iptables.go#L34)):*
Mention that arguments with embedded whitespace are not supported (because of the naive `String()` function)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 10 files reviewed at latest revision, 14 unresolved discussions.

---

*[acceptance/chaos_test.go, line 307 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPo9FKBBp92s7NqPTo:-KFPw70eGhrpbIHO3UKk:-772404593) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/chaos_test.go#L307)):*
Must've snuck in on the rebase. Removed.

---

*[acceptance/partition_test.go, line 1 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFQWeN4Re22KBzr56ay:-KFQxw8BaYlvRfTbTiQ7:1392589244) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/partition_test.go#L1)):*
Added. Didn't we have some license-checking script Pete wrote? May that's worthy for promotion to `make check` (or we just use a 'grep').

---

*[acceptance/cluster/cluster.go, line 39 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqBoj8QrcMQlpDY6q:-KFQsFAteMat7aD8tNu7:1565019282) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/cluster.go#L39)):*
It's used to get the IP address for iptables. Changed to `net.Addr`.

---

*[acceptance/cluster/localcluster.go, line 36 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPpYichEFmZFc7HrRc:-KFQywmwAPmVwQa13ZtS:-995993967) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L36)):*
I don't see how - `engine-api/client` doesn't have anything as far as I can tell: https://godoc.org/github.com/docker/engine-api/client :(

---

*[acceptance/cluster/localcluster.go, line 384 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPpq-SbEXIRs46Nxmx:-KFQtSYeVaT1HrMwRnbn:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L384)):*
Done.

---

*[acceptance/cluster/localcluster.go, line 737 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqR6ITDtz0izh-4zr:-KFQz4ZaQ-n0Lhfsl7wB:-1043362380) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L737)):*
How does that help? I need the port.

---

*[acceptance/cluster/localcluster.go, line 811 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqkSWBitVRwCbDEi0:-KFQz6Kue53XatbQ04yi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L811)):*
Done.

---

*[acceptance/cluster/localcluster.go, line 826 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqwkpuakNZq7UwZni:-KFQzQAvZ_24F4JMAmjX:-1299607173) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L826)):*
Both `%s` and `%v` do that, but that seems ok here. Changed.

---

*[acceptance/iptables/iptables.go, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBip_uwPItZGyzD_J7E:-KFQzqqzAhjFFAyDjbDx:1899379154) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/iptables/iptables.go#L15)):*
Done (I think this refers to Bidirectional below, correct?)

---

*[acceptance/iptables/iptables.go, line 15 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPrDRCAv1Wck5nf-O2:-KFQztnIelAGhsewQeSv:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/iptables/iptables.go#L15)):*
Done.

---

*[acceptance/iptables/iptables.go, line 34 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFQXrd2TeTVrvJ_PKUO:-KFR-2Z03fVHWZJYUC83:-939461467) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/iptables/iptables.go#L34)):*
Done:

> // Cmd is a naive command without proper support for whitespace.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 10 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[acceptance/partition_test.go, line 1 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFQWeN4Re22KBzr56ay:-KFR39gJ_l33d-83R9h7:-743428403) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/partition_test.go#L1)):*
I think you're talking about `build/deplicense.sh`, which attempts to determine the license for a dependency repo; it doesn't check that every file in the repo has the appropriate header. It would be easy to grep for this in `make check`.

---

*[acceptance/iptables/iptables.go, line 15 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KBip_uwPItZGyzD_J7E:-KFR3eHjvfpUzcUzOEnp:677250776) ([raw file](https://github.com/cockroachdb/cockroach/blob/4abbb6e0b1e69679fd2fdb5c343afdef437e1bee/acceptance/iptables/iptables.go#L15)):*
Yeah

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 10 files reviewed at latest revision, 13 unresolved discussions.

---

*[acceptance/partition_test.go, line 1 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFQWeN4Re22KBzr56ay:-KFR6e3oQRaFmsS-Qf9m:1211108898) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/partition_test.go#L1)):*
Ok, have a reminder in my queue.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 10 files at r8, 1 of 1 files at r9.
Review status: 8 of 10 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[acceptance/cluster/localcluster.go, line 36 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPpYichEFmZFc7HrRc:-KFR_p8kqOQiT6iTo9cu:779267018) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L36)):*
How big is that function? Could we copy pasta it and file an issue upstream?

---

*[acceptance/cluster/localcluster.go, line 737 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqR6ITDtz0izh-4zr:-KFR_xty__APQi3rHu7W:-888480976) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L737)):*
Do you? the only caller parses out the host name and throws away the port. What am I missing?

---

*[acceptance/cluster/localcluster.go, line 285 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFRaIuNOpTDp4SHh6Pk:-KFRaIuNOpTDp4SHh6Pl:994564771) ([raw file](https://github.com/cockroachdb/cockroach/blob/5174bacea9063b823c08bab94001cb4b66a4d1d4/acceptance/cluster/localcluster.go#L285)):*
man, this is weird - is it just that the API is inconsistent in using `string` instead of `NetworkMode` here and below, or are we conflating things by making `networkID`'s type `NetworkMode`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 5 of 12 files reviewed at latest revision, 8 unresolved discussions.

---

*[acceptance/cluster/localcluster.go, line 36 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPpYichEFmZFc7HrRc:-KFTK5kJCvFQiZqR7FAR:669525722) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L36)):*
It's like 100loc. I cribbed it and filed an issue (pinged this PR).

---

*[acceptance/cluster/localcluster.go, line 737 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFPqR6ITDtz0izh-4zr:-KFTMl6-uBoqlpt3MTnU:1451723162) ([raw file](https://github.com/cockroachdb/cockroach/blob/b0c3f9162ecf9f474eadf103f7ecbbf66a68d6e4/acceptance/cluster/localcluster.go#L737)):*
Oh, sorry, you're right. Changed to return `net.IP` (I do want the IP and not the host name).

---

*[acceptance/cluster/localcluster.go, line 285 \[r9\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFRaIuNOpTDp4SHh6Pk:-KFTOFUN3cYY0T8jL1At:-1378722125) ([raw file](https://github.com/cockroachdb/cockroach/blob/5174bacea9063b823c08bab94001cb4b66a4d1d4/acceptance/cluster/localcluster.go#L285)):*
It's... weird. I think `string` works better overall. Changed back.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 7 of 11 files at r1, 4 of 5 files at r2, 5 of 5 files at r3, 11 of 11 files at r4, 4 of 10 files at r5, 2 of 10 files at r8, 5 of 6 files at r10.
Review status: 10 of 12 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[acceptance/chaos_test.go, line 369 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFTiRGO49gh4W7L47RG:-KFTiRGO49gh4W7L47RH:-519441984) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d71deeffa4b1991a4d42652b10a802e5bc6c9e/acceptance/chaos_test.go#L369)):*
How come this is in this change?

aren't 3 nodes enough? also include the number in the skip message

---

*[acceptance/cluster/localcluster.go, line 66 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFThRgWMR7wRe9H4dF0:-KFThRgWMR7wRe9H4dF1:-273627707) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d71deeffa4b1991a4d42652b10a802e5bc6c9e/acceptance/cluster/localcluster.go#L66)):*
no longer used?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 8 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[acceptance/chaos_test.go, line 369 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFTiRGO49gh4W7L47RG:-KFTnbS_gNatTaNBS9VF:-1748486101) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d71deeffa4b1991a4d42652b10a802e5bc6c9e/acceptance/chaos_test.go#L369)):*
Drive-by. Fixed.

---

*[acceptance/cluster/localcluster.go, line 66 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFThRgWMR7wRe9H4dF0:-KFTnrdGMk9Q3GN6uond:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d71deeffa4b1991a4d42652b10a802e5bc6c9e/acceptance/cluster/localcluster.go#L66)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r11.
Review status: 10 of 12 files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tbg committed:
Fix `make check`

It was not correctly excluding files for os.Getenv, for example
it would complain about `os.Getenv` in `acceptance/util_test.go`
which should be excluded.

tamird commented:
Reviewed 1 of 13 files at r12, 10 of 12 files at r13.
Review status: 11 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[Makefile, line 189 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFVVtkbvH8LRhGOgzG0:-KFVVtkbvH8LRhGOgzG1:2060703881) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fdb57f917a8a5e43b482cbb32088ea76933c32/Makefile#L189)):*
did you mean to change `time\.Now` to `time.Now`? also why the inconsistent application of `-l`?

---

*[Makefile, line 191 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFVW2BqC8pOEVQjwMGm:-KFVW2BqC8pOEVQjwMGn:861456680) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fdb57f917a8a5e43b482cbb32088ea76933c32/Makefile#L191)):*
why does it need to be `$$`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tbg committed:
acceptance: Introduce partition nemesis

This nemesis allows node partitions to be set up during tests. Partitions can
be bidirectional, but also one-way (i.e. one node can initiate connections to
another, but not the other way around). The latter likely isn't too
interesting with the change to streaming RPCs, but it should be possible to
introduce interesting phenomena nevertheless.

Wrote a bank test with the nemesis. I found the existing bank test too
complicated for its own sake, so I re-wrote it. The plan is to make it
the more advanced bank example with constraints which would fail under
SNAPSHOT isolation (i.e. transferring money from A and B to C and D under
the constraint that A+B >= 0).

tschottdorf commented:
Review status: 10 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[Makefile, line 189 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFVVtkbvH8LRhGOgzG0:-KFV_c-t-EC1xuBWJebg:77090704) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fdb57f917a8a5e43b482cbb32088ea76933c32/Makefile#L189)):*
Yep. I tried using `-l` wherever possible (some of the other ones `grep` out parts from the matched line). Did I miss any?

---

*[Makefile, line 191 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFVW2BqC8pOEVQjwMGm:-KFV__1btOEUEi65xPjM:-752842657) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fdb57f917a8a5e43b482cbb32088ea76933c32/Makefile#L191)):*
Makefile :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r14.
Review status: 11 of 13 files reviewed at latest revision, 4 unresolved discussions.

---

*[Makefile, line 189 \[r13\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KFVVtkbvH8LRhGOgzG0:-KFWCvvYL47so4lVpw63:-438428932) ([raw file](https://github.com/cockroachdb/cockroach/blob/20fdb57f917a8a5e43b482cbb32088ea76933c32/Makefile#L189)):*
Nah, I think you got 'em all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 12 files at r13.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[acceptance/partition.go, line 58 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_T-PxVtr5IbDd4cEo:-KF_T-Py8ipijATDc1Mu:1528538349) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition.go#L58)):*
I get that you're trying to assert the type, but I think there's no need to make the function variable; just add a static assertion instead:
```
var _ NemesisFn = BidirectionalPartitionNemesis
```

---

*[acceptance/partition.go, line 97 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_TWzoGP1umtB20ofi:-KF_TWzoGP1umtB20ofj:2019013748) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition.go#L97)):*
what's the point of this goroutine? looks like the function blocks until it exists?

---

*[acceptance/partition_test.go, line 32 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_TgAf01opjHFoZXm3:-KF_TgAf01opjHFoZXm4:-72763320) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition_test.go#L32)):*
what does this mean?

---

*[acceptance/partition_test.go, line 52 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_TyjC5iE-YCrlXTed:-KF_TyjDUcxbcXB5jDJ8:-1220542818) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition_test.go#L52)):*
can you add a TODO to share this code with other bank test(s)?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-:-KF_U7P6fAOiG_lsnFVP:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[acceptance/partition.go, line 58 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_T-PxVtr5IbDd4cEo:-KFcJ2870YurIxa568bB:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition.go#L58)):*
Done.

---

*[acceptance/partition.go, line 97 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_TWzoGP1umtB20ofi:-KFcK0omnzC2UC5SFnfh:543278711) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition.go#L97)):*
Goroutine removed.

---

*[acceptance/partition_test.go, line 32 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_TgAf01opjHFoZXm3:-KFcL5I5RshNXu8GgRzx:-342850267) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition_test.go#L32)):*
When you're changing around the iptables stuff, this is the test you're going to want to use for manually poking around in the containers.

---

*[acceptance/partition_test.go, line 52 \[r14\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712#-KF_TyjC5iE-YCrlXTed:-KFcLcBV9XQPwG9_kARe:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0d6bbaf84c283de25ead3c538bd0b533eb1b9d0/acceptance/partition_test.go#L52)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 13 of 13 files at r16.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/4712)*
<!-- Sent from Reviewable.io -->
